### PR TITLE
Added checking markdown links

### LINKED ISSUE #80: Add markdown link checker action
## 🚀 Feature
Inspired by https://github.com/Lightning-AI/lightning/pull/16168 authored by @shenoynikhil, we can include the link checker in this repository, too, so that all our projects benefit from the GH action 🚀

### Pitch
Create a reusable workflow file `.github/workflows/check-md-links.yml` (or a better name).

### PR DESCRIPTION
**Pull request by shenoynikhil**
# Before submitting
Based on the https://github.com/Lightning-AI/lightning/pull/16168, added workflow files for checking markdown links. 

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Fixes #80.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shenoynikhil** [comment]
@akihironitta I guess I need some approval to run workflows.

File: .github/workflows/check-md-link.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+name: Check Markdown links
+# https://github.com/gaurav-nelson/github-action-markdown-link-check
+
+on:
+  push:
+    branches: [master, "release/*"]
+  pull_request:
+    branches: [master, "release/*"]
```
Comment (Borda) [inline .github/workflows/check-md-link.yml:8]:
```suggestion
    branches: [main]
  pull_request:
    branches: [main]
```

**Review by Borda — APPROVED** [review]
@akihironitta I think we shall describe in the readme that `check-` are meant to be a re-usable workflow and `ci-` to run for the particular repo

**Comment by shenoynikhil** [comment]
I think I need to update the config json to allow for skipping CHANGELOG links.

**Review by akihironitta — APPROVED** [review]
In this PR or in a follow-up PR, could you call this reusable workflow from `ci-use-checks.yml`? This way, we know the new workflow actually works :)

**Comment by akihironitta** [comment]
@shenoynikhil Oh sorry this is the correct one! https://lightning-devtools.readthedocs.io/en/latest/

@Borda We might want to update the domain to something like `lightning-utilities.readthedocs.io`. What do you think?

**Comment by shenoynikhil** [comment]
Should I use the one you have shared for now?

**Comment by shenoynikhil** [comment]
I updated the README with the link you shared. I guess whenever the domain changes, the README.md will also change accordingly.

**Review by akihironitta — CHANGES_REQUESTED** [review]
Requesting changes just for me to remind myself of coming back to this PR :)

**Comment by shenoynikhil** [comment]
Thank you so much!

**Comment by Borda** [comment]
> We might want to update the domain to something like `lightning-utilities.readthedocs.io`. What do you think?

renamed as https://lit-utilities.readthedocs.io/en/latest/

File: README.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 [![UnitTests](https://github.com/Lightning-AI/utilities/actions/workflows/ci-testing.yml/badge.svg?event=push)](https://github.com/Lightning-AI/utilities/actions/workflows/ci-testing.yml)
 [![Apply checks](https://github.com/Lightning-AI/utilities/actions/workflows/ci-use-checks.yml/badge.svg?event=push)](https://github.com/Lightning-AI/utilities/actions/workflows/ci-use-checks.yml)
-[![Documentation Status](https://readthedocs.org/projects/lightning_utilities/badge/?version=latest)](https://lightning-tools.readthedocs.io/en/latest/?badge=latest)
+[![Documentation Status](https://readthedocs.org/projects/lightning_utilities/badge/?version=latest)](https://lightning-devtools.readthedocs.io/en/latest/)
```
Comment (Borda) [inline README.md:5]:
```suggestion
[![Documentation Status](https://readthedocs.org/projects/lit-utilities/badge/?version=latest)](https://lit-utilities.readthedocs.io/en/latest/?badge=latest)
```

File: .github/workflows/check-md-links.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+    types: [opened, reopened, ready_for_review, synchronize]
+    paths:
+      - ".github/workflows/md-check-links.yml"
+      - "**/*.md"
```
Comment (akihironitta) [inline .github/workflows/check-md-links.yml:12]: Mind using a `workflow_call` trigger just like this?
https://github.com/Lightning-AI/utilities/blob/face8f06c7dd81db2991e2f8e815e0b300340081/.github/workflows/check-schema.yml#L4

(By having `workflow_call` trigger type, this `check-md-links.yml` workflow can be called from other workflows!)

Also, could you reuse this workflow in this calling workflow?
https://github.com/Lightning-AI/utilities/blob/face8f06c7dd81db2991e2f8e815e0b300340081/.github/workflows/ci-use-checks.yml

**Review by akihironitta — COMMENTED** [review]
Sorry for the delay! Thank you for working on this :)

**Comment by Borda** [comment]
hi @akihironitta @shenoynikhil what is missing to land this? :)

**Comment by akihironitta** [comment]
Hi @shenoynikhil! Mind having a look at my comment above? If you're too busy to finish this, I can contribute to this by directly pushing commits to your branch as a weekend project (if you're fine with it) 🚀

**Comment by shenoynikhil** [comment]
> Hi @shenoynikhil! Mind having a look at my comment above? If you're too busy to finish this, I can contribute to this by directly pushing commits to your branch as a weekend project (if you're fine with it) 🚀

Sure, I'll do it.

File: .github/workflows/check-md-links.yml (hunk 3 above)
Comment (shenoynikhil) [reply .github/workflows/check-md-links.yml]: Can you check now?

File: .github/workflows/check-md-links.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+jobs:
+  markdown-link-check:
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@master
+    - uses: gaurav-nelson/github-action-markdown-link-check@v1
+      with:
+        use-quiet-mode: 'yes'
+        config-file: '.github/workflows/markdown.links.config.json'
```
Comment (Borda) [inline .github/workflows/check-md-links.yml:23]: if we g with `workflow_call` then this shall be input argument
also I would consider to make this file on the fly (dump to the CI run) if none is provided :chipmunk:

File: .github/workflows/check-md-links.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+on:
+  workflow_call:
+    push:
+      branches: [main]
+    pull_request:
+      branches: [main]
+      types: [opened, reopened, ready_for_review, synchronize]
+      paths:
+        - ".github/workflows/md-check-links.yml"
+        - "**/*.md"
```
Comment (Borda) [inline .github/workflows/check-md-links.yml:13]:
```suggestion
  push:
    branches: [main]
  pull_request:
    branches: [main]
    types: [opened, reopened, ready_for_review, synchronize]
    paths:
      - ".github/workflows/md-check-links.yml"
      - "**/*.md"
```

Comment (shenoynikhil) [reply .github/workflows/check-md-links.yml]: Fixed this.

File: .github/workflows/check-md-links.yml (hunk 4 above)
Comment (shenoynikhil) [reply .github/workflows/check-md-links.yml]: That makes sense, did the first part. Will have to think about how to do the second bit. Based of initial ideas (ignore my github workflow building skills), would the jobs part of the workflow look like this,
```
jobs:
  markdown-link-check:
    runs-on: ubuntu-20.04
    steps:    
    - uses: actions/checkout@master
    - name: Check for config file, if not create it
      run: |
        if [ ! -f ${{ inputs.config-file }} ]; then
          echo "config.json not found, creating one..."
          echo '{
              "ignorePatterns": [
                {
                  "pattern": "^https://github.com/Lightning-AI/lightning/pull/.*"
                }
              ],
              "httpHeaders": [
                {
                  "urls": ["https://github.com/", "https://guides.github.com/", "https://help.github.com/", "https://docs.github.com/"],
                  "headers": {
                    "Accept-Encoding": "zstd, br, gzip, deflate"
                  }
                }
              ]
            }' > ${{ inputs.config-file }}
        else
          echo "config.json found, no need to create one."
        fi
    - uses: gaurav-nelson/github-action-markdown-link-check@v1
      with:
        use-quiet-mode: 'yes'
        config-file: ${{ inputs.config-file }}
```

**Comment by Borda** [comment]
@akihironitta ^^

**Comment by Borda** [comment]
@akihironitta ^^ :chipmunk:

**Pull request merged**